Fixes & Improvements to Renderers

twilson90 opened the pull request:
This is still a WIP, not extensively tested yet, bound to be some unexpected problems.
This isn't ready for merging yet but I will continue to work on it.
As the branch name suggests, this started as a simple fix for bitmapFills but I kept finding things that were broken and needed fixing...

CanvasGraphics:
    - Added draw triangles bitmapFill matrix support (like Flash if no UV supplied)
    - Fixed draw triangles uv mapping issues
    - Added draw triangles shared edge gap filler (inherant problem in the way Canvas draws subpixel diagonals, small gaps between shared edges. Hacky, needs work)
    - Added draw triangles stroke / lineStyle support (like Flash)
    - Tidied up scale9Grid implementation
    - Tidied up render / playCommands / hitTest
    - Improved drawRoundRect scale9Grid implementation
    - Fixed various hitTesting issues (top-left false stroke detection, scale9Grid, drawTriangles, and more probably)

CairoGraphics:
    - Added draw triangles bitmapFill matrix support
    - Fixed draw triangles uv mapping
    - Tidied up scale9Grid implementation
    - Tidied up render / playCommands / hitTest
    - Improved drawRoundRect scale9Grid implementation

Context3DGraphics:
    - Added DRAW_ELLIPSE, DRAW_CIRCLE, DRAW_ROUND_RECT

Context3DDisplayObject:
    - Fixed opaqueBackground + renderer clearing issue (now resembles Flash behaviour)

Todo:
    - scale9Grid clipping and scaling issue with strokes:
    <img width="1109" height="702" alt="chrome_kXx7L1Z795" src="https://github.com/user-attachments/assets/cf614df8-ccac-41e8-ad9a-ea38acab07ff" />
    - scale9Grid if width/height reduce by more than scale9Grid width/height then it must revert to regular scaling x/y

Commit:
added support for bitmap fill matrix in Context3DGraphics renderer.

Commit:
Fixes & Improvements to Renderers

Still a WIP, not extensively tested yet...

CanvasGraphics:
    Added draw triangles bitmapFill matrix support (like Flash if no UV supplied)
    Fixed draw triangles uv mapping issues
    Added draw triangles internal edge gap filler (for Canvas, kinda hacky, needs work)
    Added draw triangles stroke support (like Flash)
    Tidied up scale9Grid implementation
    Tidied up render / playCommands / hitTest
    Improved drawRoundRect scale9Grid implementation
    Fixed various hitTesting issues (top-left false stroke detection, scale9Grid, drawTriangles, and more probably)

CairoGraphics:
    Added draw triangles bitmapFill matrix support
    Fixed draw triangles uv mapping
    Tidied up scale9Grid implementation

Context3DGraphics:
    Added DRAW_ELLIPSE, DRAW_CIRCLE, DRAW_ROUND_RECT

Context3DDisplayObject:
    Fixed opaqueBackground + renderer clearing issue (now resembles Flash behaviour)

Todo:
    scale9Grid clipping and scaling issue with strokes.
    If width/height reduced by more than scale9Grid width/height then it should revert to regular scaling x/y like Flash.

Commit:
Fixes & Improvements to Renderers

Still a WIP, not extensively tested yet...

CanvasGraphics:
    Added draw triangles bitmapFill matrix support (like Flash if no UV supplied)
    Fixed draw triangles uv mapping issues
    Added draw triangles internal edge gap filler (for Canvas, kinda hacky, needs work)
    Added draw triangles stroke support (like Flash)
    Tidied up scale9Grid implementation
    Tidied up render / playCommands / hitTest
    Improved drawRoundRect scale9Grid implementation
    Fixed various hitTesting issues (top-left false stroke detection, scale9Grid, drawTriangles, and more probably)

CairoGraphics:
    Added draw triangles bitmapFill matrix support
    Fixed draw triangles uv mapping
    Tidied up scale9Grid implementation

Context3DGraphics:
    Added DRAW_ELLIPSE, DRAW_CIRCLE, DRAW_ROUND_RECT

Context3DDisplayObject:
    Fixed opaqueBackground + renderer clearing issue (now resembles Flash behaviour)

Todo:
    scale9Grid clipping and scaling issue with strokes.
    If width/height reduced by more than scale9Grid width/height then it should revert to regular scaling x/y like Flash.

joshtynjala commented:
Just a heads up. It’s likely that we’ll ask you to split these up into separate PRs to make it easier to test and review each thing properly.

twilson90 commented:
Looking into the scale9Grid scaling issue further, I realised Graphics.__bounds is calculated with each draw command, including the stroke thickness.
This doesn't account for scale9Grid (which might be applied after the draw commands) which might affect the thickness of strokes.
__bounds has to be calculated on demand and recalculated with a dirty flag to accomodate scale9Grid.
Something like:
`if drawCommand changes: bounds_dirty = true`
`if scale9Grid && scale changes: bounds_dirty = true`
And then each time bounds is demanded it's recalculated if dirty.
`getBounds(stroke:Bool) { if dirty return __bounds else { // calculate } }`
This would also allow for getRect to return a different result, as it's not supposed to include strokes. Currently they return the same rectangle.

Also, ignoring scale9Grid, I also noticed I get slightly different rendering results for strokes when scaled when compared to Flash:

<img width="406" height="278" alt="image" src="https://github.com/user-attachments/assets/10649d00-cbde-486e-8cdd-61d3faa9e12b" />

On the left is Flash, on the right is HTML5 Canvas. (again to emphasize, no scale9Grid)
It appears Flash averages the thickness of all strokes in quite an odd way... I have a thickness of 50, but when I set scaleX to 0.5 the overall line thickness becomes 40. Not sure how it's getting that number.

Commit:
Radial + Linear Gradient improvements

Commit:
Improved Graphics to also track the bounds excluding stroke.
Fixed getRect to return bounds excluding stroke.

Commit:
- Graphics bounds calculation overhaul: Graphics bounds are now calculated on demand when a dirty flag is set (whenever commands are added / cleared, and potentially for scale9Grid). The original method led to certain inaccuracies, for instance, the bounds of a stroked thick line with CapsStyle/JointsStyle can change depending on whether the path is closed or open, but due to the way the graphics API has no explicit end command, the end of a fill / closed paths were impossible to determine in some circumstances. Also with the new implementation we can use it for calculating fill / stroke bounds in scale9Grid, when the old method was too limited (still some work to do here). In the new implementation, JointTypes & CapsStyle are now considered, whereas before there was just an arbitrary padding for miter joints.
 - Support for DisplayObject.getRect(): New variable __boundsExStroke, which is a separate rectangle for storing the bounds of an object excluding thick strokes. Also added extra parameter to __getBounds() functions, exStroke:Bool.
 - Additional fixes to DisplayObject.opaqueBackground for all targets: Originally was not accounting for local bounds x / y. Not exactly the same as Flash now but cloose enough, Flash appears to draw individual opaqueBackgrounds for each child of a DisplayObjectContainer, unless there's overlap, then it draws a rect encompassing the 2 overlapping rects. I've found a couple strange quirks/bugs with Flash's implementation, particularly some issues with the bounds of the opaqueBackground when strokes are applied (especially mitered strokes) it can be very inaccurate. No such issues with this new implementation.
 - Path closures and various drawing API inconsistencies updated to match Flash's: Originally I was seeing minor differences with the way OpenFl handled unclosed paths with fills, manually closed paths (where lineTo/curveTo explicitly closes the path), or the way it handles beginFill in the middle of a series of lineTo/curveTo commands. I have rewritten it for Canvas and Cairo so it now matches Flash's behaviour.
 - Fixed Gradients for Canvas / Cairo: After struggling to get Gradients working in the Canvas Renderer I took a look at Ruffle and lifted their solution. There's still an issue with stroke gradients which I can't see a reasonable fix for.
 - Tidied up the Scale9Grid implementation some more.
 - Removed what appears to be an arbitrary 1.5px pad in OpenGLRenderer.__scissorRect.

 TODO:
 - Fix scale9Grid objects with strokes
 - Thoroughly test Graphics API.
 - Thoroughly test scale9Grid implementation.

twilson90 commented:
Updated the branch:

-  Graphics bounds calculation overhaul: Graphics bounds are now calculated on demand when a dirty flag is set (whenever commands are added / cleared, and potentially when scale9Grid is set). The original method led to certain inaccuracies, for instance, the bounds of a stroked thick line with CapsStyle/JointsStyle can change depending on whether the path is closed or open, but due to the way the graphics API will automatically resolve paths on render, the end of a fill / closed paths were impossible to determine in this incremental way. Also with the new implementation we can use it for calculating fill / stroke bounds in scale9Grid, when the old method was too limited (still some work to do here). In the new implementation, JointTypes & CapsStyle are now considered, whereas before there was just an arbitrary padding for miter joints.
- Support for DisplayObject.getRect(): New variable __boundsExStroke, which is a separate rectangle for storing the bounds of an object excluding thick strokes. Also added extra parameter to __getBounds() functions, exStroke:Bool.
- Additional fixes to DisplayObject.opaqueBackground for all targets: Originally was not accounting for local bounds x / y. Not exactly the same as Flash now but cloose enough, Flash appears to draw individual opaqueBackgrounds for each child of a DisplayObjectContainer, unless there's overlap, then it draws a rect encompassing the 2 overlapping rects. I've found a couple strange quirks/bugs with Flash's implementation, particularly some issues with the bounds of the opaqueBackground when strokes are applied (especially mitered strokes) it can be very inaccurate. No such issues with this new implementation.
- Path closures and various drawing API inconsistencies updated to match Flash's: Originally I was seeing minor differences with the way OpenFl handled unclosed paths with fills, manually closed paths (where lineTo/curveTo explicitly closes the path), or the way it handles beginFill in the middle of a series of lineTo/curveTo commands. I have rewritten it for Canvas and Cairo so it now matches Flash's behaviour.
- Fixed Gradients for Canvas / Cairo: After struggling to get Gradients working in the Canvas Renderer I took a look at Ruffle and lifted their solution. There's still an issue with stroke gradients which I can't see a reasonable fix for.
- Tidied up the Scale9Grid implementation some more.
- Removed what appears to be an arbitrary 1.5px pad in OpenGLRenderer.__scissorRect.

TODO:
- Fix scale9Grid objects with strokes
- Thoroughly test Graphics API.
- Thoroughly test scale9Grid implementation.

Commit:
- DisplayObject.scale9Grid implementation simplified + finished. Graphics made dirty when scale9Grid set, Scale9Grid.hx + Scale9GridBounds.hx deleted and merged into Graphics.hx.
- DisplayObject.opaqueBackground further improvements + fixes. DisplayObject.cacheAsBitmap now render opaqueBackground. GL renders rotated rect now. Also fixed issue with software render on Canvas + Cairo was also rendering the opaqueBackground to __cacheBitmap in DisplayObject.hx.
- ColorTransform.__isDeafult and ColorTransform.__equals altered to allow for very small differences in multipliers. I noticed unnecessary BitmapData.colorTransform() calls when multipliers = 0.99999, which happens after an identity ColorTransform is inverted and combined in the render pipeline.
- Basic LineScaleMode support (not identical to Flash due to inherant canvas limitations but sort of half-way between openFl and Flash).
- Added Graphics.beginShaderFill matrix parameter support.
- Added Graphics.__readGraphicsData support for LINE_GRADIENT_STYLE, LINE_BITMAP_STYLE + BEGIN_SHADER_FILL.
- Added scale9Grid support for DRAW_QUADS in CairoGraphics & CanvasGraphics.
- Simplified CairoGraphics & CanvasGraphics hitTest function.

- Fixed Graphics bounds calculation.
- Fixed CairoGraphics & CanvasGraphics hitTest.
- Fixed CanvasTextField not rendering.

Commit:
Replaced null-coalescing operators (??) for backwards compatibility

twilson90 commented:
- DisplayObject.scale9Grid implementation simplified + finished. Graphics made dirty when scale9Grid set, Scale9Grid.hx + Scale9GridBounds.hx deleted and merged into Graphics.hx. Stroke thicknesses taken into account.
- DisplayObject.opaqueBackground further improvements + fixes. DisplayObject.cacheAsBitmap now render opaqueBackground. GL renders rotated rect now. Also fixed issue with software render on Canvas + Cairo was also rendering the opaqueBackground to __cacheBitmap in DisplayObject.hx.
- ColorTransform.__isDefault and ColorTransform.__equals altered to allow for very small differences in multipliers. I noticed unnecessary BitmapData.colorTransform() calls when multipliers = 0.99999, which happens after an identity ColorTransform is inverted and combined in the render pipeline.
- Added basic LineScaleMode support (not identical to Flash due to inherant canvas limitations but sort of half-way between openFl and Flash).
- Added Graphics.beginShaderFill matrix parameter support.
- Added Graphics.__readGraphicsData support for LINE_GRADIENT_STYLE, LINE_BITMAP_STYLE + BEGIN_SHADER_FILL.
- Added scale9Grid support for DRAW_QUADS in CairoGraphics & CanvasGraphics.
- Simplified CairoGraphics & CanvasGraphics hitTest function.

Fixes:
- Fixed Graphics bounds calculation.
- Fixed CairoGraphics & CanvasGraphics hitTest.
- Fixed CanvasTextField not rendering.

Test suite demonstrating all the old issues with openFL that I've addressed:
https://github.com/twilson90/openfl-test-suite

Try building in current version of openFl (or dev branch) to see all the original issues as they were, and build again in this branch to see all the fixes.

Commit:
- Fixed Scale9Grid offset calculation
- Fixed Scale9Grid pattern matrix calculation
- Unified mask drawing logic allowing for complex masks.
- Fixed hitTest for masks.
- Fixes for some edge-case drawing sequences involving Graphics.lineStyle. Results now consistent on all targets with Flash.
- Rearranged code for consistency between CanvasGraphics.hx + CairoGraphics.hx

Commit:
Fixed negatively scaled ellipse / round rect in Context3D returning minimum number of vertices.

Commit:
- Removed unnecessary variables
- Removed optimization variant for DRAW_RECT command in canvas, did't see any benefit to it.
- Various checks to prevent draw calls that effectively draw nothing
- fixed drawTriangles being painted twice

Commit:
Merge remote-tracking branch 'upstream/develop' into gl-bitmap-fill-matrix

Commit:
- Fixed issues with graphics command processing.
- Refactored CanvasGraphics & CairoGraphics further to make them close to identical in structure.
- Fixed multiple issues with scale9Grid: For more complex Graphics with multiple shapes, contents were not scaling correctly and bitmap fill alignment was wrong. Also pixelRatio was ignored.
- Fixed Graphics.drawQuads positioning. For some reason the position was influenced by the last moveTo or lineTo position, removed that because it seemed wrong / counter-intuitive.
- Fixed Graphics.drawQuads hitTest.
- Fixed error states in cairo due to invalid matrices.
- Re-implemented Graphics.drawRect optimization in canvas and cairo.

Also a side effect of previous changes, strokes  are now aligned for pixel perfect rendering if the bounds of the displayObject and the worldTransform have integer translations. This is due to the more accurate bounds calculation in Graphics.hx and the way it's applied in the rounding of the __renderTransform subpixel offset.

Commit:
- Fixed issues with graphics command processing.
- Refactored CanvasGraphics & CairoGraphics further to make them close to identical in structure.
- Fixed multiple issues with scale9Grid: For more complex Graphics with multiple shapes, contents were not scaling correctly and bitmap fill alignment was wrong.
- Fixed Graphics.drawQuads positioning. For some reason the position was influenced by the last moveTo or lineTo position, removed that because it seemed wrong / counter-intuitive.
- Fixed Graphics.drawQuads hitTest.
- Fixed error states in cairo due to invalid matrices.
- Re-implemented Graphics.drawRect optimization in canvas and cairo.

Also a side effect of previous changes, strokes  are now aligned for pixel perfect rendering if the bounds of the displayObject and the worldTransform have integer translations. This is due to the more accurate bounds calculation in Graphics.hx and the way it's applied in the rounding of the __renderTransform subpixel offset.

Commit:
Merge branch 'gl-bitmap-fill-matrix' of https://github.com/twilson90/openfl into gl-bitmap-fill-matrix

Commit:
fix Graphics.__hitTest false positive

Commit:
removed comma

Commit:
use IntMap<Bool> instead of Map<Int,Bool>

Commit:
Fix seenEdgeMap.clear() test fail + remove useless stroke parameters

Commit:
ci: rerun

Commit:
Further efforts to replicate Flash's behaviour in Graphics commands processing, like implicit closing of paths and the order of fills/strokes. This has its limits though, as I've discovered some issues with Flash's Graphics API that can lead to some kind of corruption, which we do not want to replicate.

Commit:
Some refactoring, simplification, moved code.
Also added hitTestCairo singleton, which is  significantly more efficient when hitTesting DisplayObjects that are not added to the hierarchy.

twilson90 commented:
I think this branch is at a stage now where it should be looked over and tested by someone else, there might be something glaring that I've missed.
When I started using openFl again towards the end of last year I was a bit dismayed by all the issues I immediately encountered... this goes some way towards fixing them.
In the process I've completely gutted the CanvasGraphics and CairoGraphics files, rewritten most of the scale9grid stuff and made the Graphics bounds calculation a lot more accurate (and fixing getRect to return stroke-excluded bounds).
The improved bounds calculation has also meant the subpixel precision of strokes is better.
Theres a load of other little fixes in there which I've tried to list in the commit messages but I've probably missed a few.
It'll be difficult to unstitch this and make into separate PRs, but I suppose it's not impossible.

Again, to see the various fixes I've also updated https://github.com/twilson90/openfl-test-suite
The issues are obvious when you compile openfl-test-suite with the latest version of openFl and contrast/compare when compiled with my branch.

Commit:
Fixes issue with multi-path masks not rendering.

Commit:
Improve hardware rendering compatibility and scale9Grid handling

- Fix Context3DGraphics scale9Grid implementation
- Replace Context3DGraphics.isCompatible check with Graphics.__isHardwareCompatible property
- Add Graphics.__isHardwareDrawable flag to track when graphics can use hardware rendering
- Allow for scale9Grid to be rendered in Context3DGraphics
- Fix Graphics.__generateUV to properly handle scale9Grid
- Add __invalidateVertexBufferOnTransform flag for necessary drawing commands to fix issue with circle/ellipse/roundRect rendering in Context3DGraphics (when transform scale changes the number of vertices on rounded shapes may change, so we set Graphics.__hardwareDirty = true)
- Remove redundant DRAW_RECT rendering logic in Context3DGraphics
- Limit maximum vertices for ellipse and roundRect rendering to prevent very large rounded shapes from producing insane amounts of vertices

Commit:
- Batch GL draw calls for sequences of Graphics draw commands that use the same fill color / bitmap.
- Converted some inline functions to static class inline functions in Context3DGraphics.hx, along with some significant refactoring.
- Minor efficiency improvements to Context3DGraphics.hx
- Fix drawCircle preparing scale applied vertices, but rendering without scale applied (vertex count mismatch leading to corrupted geometry)
- Fix condition to skip children in DisplayObjectContainer.__getBounds when both scaleX or scaleY are zero. Leads to inconsistent bounds calculation between child and parent if child is 0 scale on a single axis.
- Added additional GL stats draw call counter (guarded by #if gl_stats) in Graphics

Commit:
Removed trace statement

Commit:
fix(display): correct stroke padding calculation for scaled scale9Grid graphics

Scale9Grid object with lineStyle & scaleX or scaleY equal to zero was resulting in infinity rectangles.

Commit:
pending review

Commit:
- Changes to the way DisplayObject.scrollRect affects bounds and worldTransform so non-Flash targets now resembles Flash's behaviour.
- Added __localBounds and __localBoundsDirty variables to DisplayObject.hx for caching local bounds. Prevents unnecessary calculation when querying the width and height.
- Implemented opaqueBackground for DOMRenderer.
- Changed opaqueBackground implementation slightly for Context3DRenderer / Context3DDisplayObject, fixing masks.
- Changed bounds calculations in DisplayObject.hx, DisplayObjectContainer.hx and SimpleButton.hx. Mirroring the code so it looks structurally similar.
- Removed obsolete moveTo(0,0) in Graphics constructor
- Allow rendering of Graphics less than 1px width or height.
- Simplified CanvasGraphics.hx and CairoGraphics.hx.
- Removed unnecessary context.closePath() in CanvasRenderer.__pushMask()
- Moved __bitmapScaleX + __bitmapScaleY assignment to Graphics.__update();
- Applied color transform fix to Context3DGraphics.hx
- Scale9Grid conditional tweaks, additional openfl_legacy_scale9grid checks.
- Added hasFillMatrix, hasStrokeMatrix booleans to CanvasGraphics.hx, CairoGraphics.hx. Skips unnecessary potentially costly applyPatternMatrix if false.
- Rmoved DOMRenderer.applyStyle + DOMRenderer.clearStyle

Commit:
Graphics.copyFrom fix
